File: 0146. LRU Cache/146.lru-cache.java
//Design a data structure that follows the constraints of a Least Recently Used
//(LRU) cache.
//
// Implement the LRUCache class:
//
//
// LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
//
// int get(int key) Return the value of the key if the key exists, otherwise
//return -1.
// void put(int key, int value) Update the value of the key if the key exists.
//Otherwise, add the key-value pair to the cache. If the number of keys exceeds
//the capacity from this operation, evict the least recently used key.
//
//
// The functions get and put must each run in O(1) average time complexity.
//
//
// Example 1:
//
//
//Input
//["LRUCache", "put", "put", "get", "put", "get", "put", "get", "get", "get"]
//[[2], [1, 1], [2, 2], [1], [3, 3], [2], [4, 4], [1], [3], [4]]
//Output
//[null, null, null, 1, null, -1, null, -1, 3, 4]
//
//Explanation
//LRUCache lRUCache = new LRUCache(2);
//lRUCache.put(1, 1); // cache is {1=1}
//lRUCache.put(2, 2); // cache is {1=1, 2=2}
//lRUCache.get(1);    // return 1
//lRUCache.put(3, 3); // LRU key was 2, evicts key 2, cache is {1=1, 3=3}
//lRUCache.get(2);    // returns -1 (not found)
//lRUCache.put(4, 4); // LRU key was 1, evicts key 1, cache is {4=4, 3=3}
//lRUCache.get(1);    // return -1 (not found)
//lRUCache.get(3);    // return 3
//lRUCache.get(4);    // return 4
//
//
//
// Constraints:
//
//
// 1 <= capacity <= 3000
// 0 <= key <= 10⁴
// 0 <= value <= 10⁵
// At most 2 * 10⁵ calls will be made to get and put.
//
//
// Related Topics Hash Table Linked List Design Doubly-Linked List 👍 21580 👎 1
//120

import java.util.LinkedHashMap;
import java.util.Map;

//leetcode submit region begin(Prohibit modification and deletion)
class LRUCache {

    private final int capacity;
    private final Map<Integer, Integer> cache = new LinkedHashMap<>();

    public LRUCache(int capacity) {
        this.capacity = capacity;
    }

    public int get(int key) {
        if (!cache.containsKey(key)) {
            return -1;
        }
        makeRecently(key);
        return cache.get(key);
    }

    public void put(int key, int val) {
        if (cache.containsKey(key)) {
            cache.put(key, val);
            makeRecently(key);
            return;
        }

        if (cache.size() >= this.capacity) {
            int oldestKey = cache.keySet().iterator().next();
            cache.remove(oldestKey);
        }
        cache.put(key, val);
    }

    private void makeRecently(int key) {
        int val = cache.get(key);
        cache.remove(key);
        cache.put(key, val);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
//leetcode submit region end(Prohibit modification and deletion)

File: 0146. LRU Cache/146.lru-cache.go
package leetcode

import "container/list"

/*
 * @lc app=leetcode id=146 lang=golang
 *
 * [146] LRU Cache
 */

// @lc code=start
type LRUCache struct {
	cap     int
	list    *list.List
	hashmap map[int]*list.Element
}

type KV struct {
	key   int
	value int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		list:    list.New(),
		hashmap: make(map[int]*list.Element),
		cap:     capacity,
	}
}

func (this *LRUCache) Get(key int) int {
	value, existed := this.hashmap[key]
	if !existed {
		return -1
	}

	this.moveExistedKeyToEnd(key)
	return value.Value.(KV).value
}

func (this *LRUCache) Put(key int, value int) {
	element, existed := this.hashmap[key]
	if existed {
		element.Value = KV{key, value}
		this.moveExistedKeyToEnd(key)
		return
	}

	newElement := this.list.PushBack(KV{key, value})
	this.hashmap[key] = newElement

	if len(this.hashmap) > this.cap {
		e := this.list.Front()
		delete(this.hashmap, e.Value.(KV).key)
		this.list.Remove(e)
	}
}

func (this *LRUCache) moveExistedKeyToEnd(key int) {
	var e *list.Element = this.hashmap[key]
	this.list.MoveToBack(e)
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
// @lc code=end
